test/fixtures/permission-validation/setup.js

```diff
@@ -0,0 +1,488 @@
+    { Category: 'Document', Catalog: 'C1', Access: new Uint64(0b1000) },
+    {
+      Category: 'PersonDocument',
+      Subsystem: 'S2',
+      Access: new Uint64(0b1000),
+    },
+    {
+      Category: 'PersonDocument',
+      Catalog: 'C1',
+      Access: new Uint64(0b1000),
+    },
+  ],
+  Blocked: [
+    { Category: 'Catalog', Access: new Uint64(0b11111) },
+    { Category: 'Category', Access: new Uint64(0b11111) },
+    { Category: 'Role', Subsystem: 'S1', Access: new Uint64(0b11111) },
+    { Category: 'Role', Subsystem: 'S2', Access: new Uint64(0b11111) },
+    { Category: 'SystemUser', Subsystem: 'S1', Access: new Uint64(0b11111) },
+    { Category: 'SystemUser', Subsystem: 'S1', Access: new Uint64(0b11111) },
+    { Category: 'Person', Subsystem: 'S1', Access: new Uint64(0b11111) },
+    { Category: 'Document', Catalog: 'C1', Access: new Uint64(0b11111) },
+  ],
+};
+
+const rawData = {
+  Document: [
+    { Name: 'documentC1', Catalog: 'C1' },
+    { Name: 'document2C1', Catalog: 'C1' },
+    { Name: 'documentC2', Catalog: 'C2' },
+  ],
+  Person: [
+    { Name: 'personS1', Subsystem: 'S1' },
+    { Name: 'person2S1', Subsystem: 'S1' },
+    { Name: 'personS2', Subsystem: 'S2' },
+  ],
+  PersonDocument: [
+    { Name: 'personDocumentC1S1', Catalog: 'C1', Subsystem: 'S1' },
+    { Name: 'personDocumentC1S2', Catalog: 'C1', Subsystem: 'S2' },
+    { Name: 'personDocumentC2S2', Catalog: 'C2', Subsystem: 'S2' },
+  ],
+};
+
+const preparePgProvider = async () => {
+  const systemSchemasDir = join(__dirname, '../../..', 'schemas', 'system');
+  const testSchemasDir = join(__dirname, 'schemas');
+  const context = {
+    api: { common: { Uint64 }, net: { isIP } },
+    Auth: { passwordStrength: password => password.length > 5 },
+  };
+  const schema = await metaschema.fs.load(
+    [systemSchemasDir, testSchemasDir],
+    { ...options, context },
+    config
+  );
+  const providerOptions = {
+    ...pgOptions,
+    schema,
+    database: 'test_permission_validation',
+  };
+
+  await pgProvider.open(providerOptions);
+  await pgProvider.setup();
+};
+
+const prepareSubsystems = async () => {
+  const subsystems = {};
+
+  await Promise.all(
+    rawSubsystems.map(async subsystem => {
+      subsystems[subsystem.Name] = await pgProvider.create(
+        'Subsystem',
+        subsystem
+      );
+    })
+  );
+
+  return subsystems;
+};
+
+const prepareRoles = async subsystems => {
+  const roles = {};
+  const createSubsystemRoles = async subsystem => {
+    const subsystemRoles = {};
+
+    await Promise.all(
+      rawRoles.map(async role => {
+        const { Name: name } = role;
+        role = {
+          ...role,
+          Name: name + subsystem,
+          Subsystem: subsystems[subsystem],
+        };
+        subsystemRoles[name] = await pgProvider.create('Role', role);
+      })
+    );
+
+    roles[subsystem] = subsystemRoles;
+  };
+
+  await Promise.all(subsystemNames.map(createSubsystemRoles));
```

Comment:
I think you can just assign `roles[subsystem] = subsystemRoles;` at the top of `createSubsystemRoles` and then return `rawRoles.map(async role => { ... })` without `await`ing it and then `await` all of them after flattening the resulting array.
Feel free to ping me if further explanation needed.

Reply:
@lundibundi, PTAL